tests/0170-share_consumer_subscription.c

```diff
@@ -0,0 +1,530 @@
+
+        TEST_CALL_ERR__(rd_kafka_subscription(rk, &cur));
+        TEST_ASSERT(cur != NULL, "subscription() returned NULL list");
+        cnt = cur->cnt;
+        rd_kafka_topic_partition_list_destroy(cur);
+        return cnt;
+}
+
+/**
+ * Consume messages, return count of valid messages received.
+ */
+static int consume_batch(rd_kafka_t *rk, int timeout_ms) {
+        rd_kafka_message_t *batch[BATCH_SIZE];
+        rd_kafka_error_t *err;
+        size_t rcvd = 0;
+        int valid = 0;
+        size_t i;
+
+        err = rd_kafka_share_consume_batch(rk, timeout_ms, batch, &rcvd);
+        if (err) {
+                rd_kafka_error_destroy(err);
+                return 0;
+        }
+
+        for (i = 0; i < rcvd; i++) {
+                if (!batch[i]->err)
+                        valid++;
+                rd_kafka_message_destroy(batch[i]);
+        }
+        return valid;
+}
+
+/**
+ * Consume until expected count reached or max attempts exhausted.
+ */
+static int consume_until(rd_kafka_t *rk, int expected, int max_attempts) {
+        int total = 0;
+
+        while (total < expected && max_attempts-- > 0)
+                total += consume_batch(rk, 2000);
+
+        return total;
+}
+
+/**
+ * Helper: Set group config to earliest offset.
+ */
+static void set_group_offset_earliest(rd_kafka_t *rk, const char *group) {
+        const char *cfg[] = {"share.auto.offset.reset", "SET", "earliest"};
+        test_IncrementalAlterConfigs_simple(rk, RD_KAFKA_RESOURCE_GROUP, group, cfg, 1);
+}
+
+
+/*
+ * Verify rd_kafka_subscription() correctly reflects subscribed topics
+ * before and after unsubscribe.
+ */
+static void do_test_subscription_count(void) {
+        const char *group = "share-sub-query";
+        const char *topics[] = {
+                "0170-share-query-t1",
+                "0170-share-query-t2",
+                "0170-share-query-t3"
+        };
+        const int topic_cnt = 3;
+        rd_kafka_t *rk;
+        int cnt, i;
+
+        TEST_SAY("Subscription introspection test\n");
+
+        for (i = 0; i < topic_cnt; i++)
+                test_create_topic_wait_exists(NULL, topics[i], 1, -1, 30000);
+
+        rk = create_share_consumer(group);
+        subscribe_topics(rk, topics, topic_cnt);
+
+        /* Verify subscription shows all topics */
+        cnt = get_subscription_count(rk);
+        TEST_ASSERT(cnt == topic_cnt, "expected %d subscriptions, got %d",
+                    topic_cnt, cnt);
+
+        /* Unsubscribe and verify empty */
+        TEST_CALL_ERR__(rd_kafka_unsubscribe(rk));
+        cnt = get_subscription_count(rk);
+        TEST_ASSERT(cnt == 0, "expected 0 after unsubscribe, got %d", cnt);
+
+        rd_kafka_consumer_close(rk);
+        rd_kafka_destroy(rk);
+
+        TEST_SAY("Subscription introspection: PASSED\n");
+}
+
+
+/*
+ * Calling unsubscribe multiple times should not error.
+ */
+static void do_test_multiple_unsubscribe(void) {
+        const char *group = "share-unsub-idem";
```

Comment:
Create this as well from the random name generator right now same `test_mk_topic_name`